File: Problems/Algorithms/1461. Check If a String Contains All Binary Codes of Size K/CheckBinaryCodes.kt
class Solution {
    fun hasAllCodes(s: String, k: Int): Boolean {
        var total = (1 shl k)
        val hashes = IntArray(total) { 0 }
        
        var hash = 0
        for (i in 0..s.length-1) {
            hash = ((hash shl 1) and ((1 shl k) - 1) or (s.get(i) - '0'))
            
            if (i >= k - 1 && hashes[hash] == 0) {
                hashes[hash] = 1
                total--
            }
        }
        
        return total == 0
    }
}

File: Problems/Algorithms/1461. Check If a String Contains All Binary Codes of Size K/check_binary_codes.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        substrings = set()

        for i in range(len(s)-k+1):
            substrings.add(s[i:i+k])

        return len(substrings) == 1 << k

